src/core/builder.rs

```diff
@@ -0,0 +1,534 @@
+        &self.first_messages[idx]
+    }
+    fn second_message(&mut self, idx: usize) -> &SecondReduceMessage<G1, G2, GT> {
+        &self.second_messages[idx]
+    }
+
+    fn process_first_reduce_message(
+        &mut self,
+        m: &FirstReduceMessage<G1, G2, GT>,
+    ) -> FirstReduceChallenge<Scalar> {
+        self.transcript.append_group(b"d1_left", &m.d1_left);
+        self.transcript.append_group(b"d1_right", &m.d1_right);
+        self.transcript.append_group(b"d2_left", &m.d2_left);
+        self.transcript.append_group(b"d2_right", &m.d2_right);
+        self.transcript.append_group(b"e1_beta", &m.e1_beta);
+        self.transcript.append_group(b"e2_beta", &m.e2_beta);
+
+        let beta = self.transcript.challenge_scalar(b"first_reduce_beta");
+        let beta_inv = beta.inv().unwrap();
+        FirstReduceChallenge {
+            beta,
+            beta_inverse: beta_inv,
+        }
+    }
+
+    fn process_second_reduce_message(
+        &mut self,
+        m: &SecondReduceMessage<G1, G2, GT>,
+    ) -> SecondReduceChallenge<Scalar> {
+        self.transcript.append_group(b"c_plus", &m.c_plus);
+        self.transcript.append_group(b"c_minus", &m.c_minus);
+        self.transcript.append_group(b"e1_plus", &m.e1_plus);
+        self.transcript.append_group(b"e1_minus", &m.e1_minus);
+        self.transcript.append_group(b"e2_plus", &m.e2_plus);
+        self.transcript.append_group(b"e2_minus", &m.e2_minus);
+
+        let alpha = self.transcript.challenge_scalar(b"second_reduce_alpha");
+        let alpha_inv = alpha.inv().unwrap();
+        SecondReduceChallenge {
+            alpha,
+            alpha_inverse: alpha_inv,
+        }
+    }
+
+    fn challenge_fold_scalars(&mut self) -> FoldScalarsChallenge<Scalar> {
+        let gamma = self.transcript.challenge_scalar(b"fold_scalars_gamma");
+        let gamma_inv = gamma.inv().unwrap();
+        FoldScalarsChallenge {
+            gamma,
+            gamma_inverse: gamma_inv,
+        }
+    }
+
+    fn challenge_scalar_product_scalars(&mut self) -> ScalarProductChallenge<Self::Scalar> {
+        let d = self.transcript.challenge_scalar(b"scalar_product_d");
+        let d_inv = d.inv().unwrap();
+        ScalarProductChallenge {
+            d,
+            d_inverse: d_inv,
+        }
+    }
+
+    fn process_scalar_product_message(&self) -> &ScalarProductMessage<G1, G2> {
+        &self.scalar_msg
+    }
+
+    fn process_vmv_message(&mut self) -> VMVMessage<G1, GT> {
+        let message = self
+            .vmv_msg
+            .as_ref()
+            .expect("VMV message must be present in verify builder");
+        self.transcript.append_group(b"c_eval_vmv", &message.c);
+        self.transcript.append_group(b"d2_eval_vmv", &message.d2);
+        self.transcript.append_group(b"e1_eval_vmv", &message.e1);
+        message.clone()
+    }
+}
+
+/// Additional debug helpers:
+impl<G1, G2, GT, Scalar, T> DoryProofBuilder<G1, G2, GT, Scalar, T>
```

Comment:
Doing a proper
```suggestion
impl<G1, G2, GT, Scalar, T> Debug for DoryProofBuilder<G1, G2, GT, Scalar, T>
```
would be a lot cleaner and would give a lot of other stuff for free.

Might make sense to create an GitHub issue for it if you want it to get done but don't want to do it right now.

As it stands, I don't think this belongs in this PR.